app/scripts/lib/multichain-api/wallet-requestPermissions.js

```diff
@@ -65,119 +68,125 @@ async function requestPermissionsImplementation(
   const [requestedPermissions] = params;
   delete requestedPermissions[Caip25EndowmentPermissionName];
 
-  const [_grantedPermissions] = await requestPermissionsForOrigin(
-    requestedPermissions,
-  );
-
-  // caveat values are frozen and must be cloned before modified
-  const grantedPermissions = { ..._grantedPermissions };
+  const legacyRequestedPermissions = pick(requestedPermissions, [
+    RestrictedMethods.eth_accounts,
+    PermissionNames.permittedChains,
+  ]);
+  delete requestedPermissions[RestrictedMethods.eth_accounts];
+  delete requestedPermissions[PermissionNames.permittedChains];
+
+  let legacyApproval;
+  if (Object.keys(legacyRequestedPermissions).length > 0) {
+    if (!legacyRequestedPermissions[RestrictedMethods.eth_accounts]) {
+      legacyRequestedPermissions[RestrictedMethods.eth_accounts] = {};
+    }
 
-  const ethAccountsPermission =
-    grantedPermissions[RestrictedMethods.eth_accounts];
+    if (!legacyRequestedPermissions[PermissionNames.permittedChains]) {
+      const { chainId } =
+        getNetworkConfigurationByNetworkClientId(networkClientId);
+      legacyRequestedPermissions[PermissionNames.permittedChains] = {
+        caveats: [
+          {
+            type: CaveatTypes.restrictNetworkSwitching,
+            value: [chainId],
+          },
+        ],
+      };
+    }
 
-  if (process.env.BARAD_DUR && ethAccountsPermission) {
-    // TODO: Use permittedChains permission returned from requestPermissionsForOrigin() when available
-    const { chainId } = getNetworkConfigurationByNetworkClientId(
-      req.networkClientId,
+    legacyApproval = await requestPermissionApprovalForOrigin(
+      legacyRequestedPermissions,
     );
+  }
 
-    const scopeString = `eip155:${parseInt(chainId, 16)}`;
-
-    const ethAccounts = ethAccountsPermission.caveats[0].value;
-
-    const caipAccounts = ethAccounts.map(
-      (account) => `${scopeString}:${account}`,
+  let grantedPermissions = {};
+  if (
+    (Object.keys(requestedPermissions).length === 0 && !legacyApproval) ||
+    Object.keys(requestedPermissions).length > 0
+  ) {
+    const [_grantedPermissions] = await requestPermissionsForOrigin(
+      requestedPermissions,
     );
+    // permissions are frozen and must be cloned before modified
+    grantedPermissions = { ..._grantedPermissions };
+  }
```

Comment:
I feel like some inline comments might help navigate this since the two different cases here (where permissions are specified and when they are not) is a bit tricky to follow as is.

Reply:
made this a bit clearer with a new variable name and inline comments https://github.com/MetaMask/metamask-extension/pull/26483/commits/f01a8a619e0d5c4fd9a3f56281aa62d4e13efe61

Reply:
Awesome thanks Jiexi!